src/frequenz/sdk/timeseries/_moving_window.py

```diff
@@ -0,0 +1,182 @@
+    # starting at 1.1.23 with samplerate=1
+    window = MovingWindow(size = (60 * 60 * 24 * 2), sample_receiver)
+
+    # wait for one full day until the buffer is filled
+    asyncio.sleep(60*60*24)
+
+    # create a polars series with one full day of data
+    time_start = datetime(2023, 1, 1)
+    time_end = datetime(2023, 1, 2)
+    s = pl.Series("Jan_1", mv[time_start:time_end])
+    ```
+    """
+
+    def __init__(
+        self,
+        size: int,
+        resampled_data_recv: Receiver[Sample],
+        sampling_period: timedelta,
+        window_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """
+        Initialize the MovingWindow.
+
+        This method creates the underlying ringbuffer and starts a
+        new task that updates the ringbuffer with new incoming samples.
+        The task stops running only if the channel receiver is closed.
+
+        Args:
+            size: The number of elements that are stored.
+            resampled_data_recv: A receiver that delivers samples with a
+                given sampling period.
+            sampling_period: The sampling period.
+            window_alignment: A datatime object that defines a point in time to which
+                the window is aligned to modulo window size.
+                (default is midnight 01.01.01)
+
+        Raises:
+            asyncio.CancelledError: when the task gets cancelled.
+        """
+        self._resampled_data_recv = resampled_data_recv
+        self._buffer = OrderedRingBuffer(
+            np.empty(shape=size, dtype=float),
+            sampling_period=sampling_period,
+            time_index_alignment=window_alignment,
+        )
+        self._copy_buffer = False
+        try:
+            _t = asyncio.create_task(self._run_impl())
+            # make the linter happy
+            assert isinstance(_t, asyncio.Task)
+        except asyncio.CancelledError:
+            log.debug("Cancelling MovingWindow task: %s", __name__)
+            raise
+
+    async def _run_impl(self) -> None:
+        """Awaits samples from the receiver and updates the underlying ringbuffer."""
+        async for sample in self._resampled_data_recv:
+            log.debug("Received new sample: %s", sample)
+            log.debug("Updating ring buffer with latest sample")
+
+            # missing values coming from the logical meter are marked as NaN
+            missing = False
+            assert isinstance(sample.value, float)
+            if isnan(sample.value):
+                missing = True
+
+            self._buffer.update(sample.timestamp, sample.value, missing)
+
+        log.error("Channel has been closed")
+
+    def __len__(self) -> int:
+        """
+        Return the size of the `MovingWindow`s underlying buffer.
+
+        Returns:
+            The size of the `MovingWindow`.
+        """
+        return len(self._buffer)
+
+    def __getitem__(self, key: (int | datetime) | slice) -> float | ArrayLike:
```

Comment:
You might need to use `Union` here instead as only parenthesis are supported in the latest versions of python/typing

Reply:
You can use `from __future__ import annotations` to get this to work in python 3.8.

Reply:
I guess I sacrificed speed for usability. But I'd say it's fine in that case. Accessing such a window shouldn't happen too often.

Reply:
> You can use from __future__ import annotations to get this to work in python 3.8.

That doesn't seem to work unfortunately. Will do the long version.

Reply:
I wonder why.  We have that kind of notation in the sdk already,  and with `from __future__ import annotations` it works just fine.

- https://github.com/frequenz-floss/frequenz-sdk-python/blob/63411e694935b4380719b85e1405bfaf8c1990b5/src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py#L424
- https://github.com/frequenz-floss/frequenz-sdk-python/blob/63411e694935b4380719b85e1405bfaf8c1990b5/src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py#L542

Reply:
btw,  in `(int | datetime) | slice`, do you need the brackets? can you try without those?

Reply:
No, brackets are not needed.